skfda/representation/grid.py

```diff
@@ -1101,30 +1105,61 @@ def restrict(
-        for dr, grid_points in zip(
-            domain_range,
-            self.grid_points,
-        ):
-            keep_index = (
-                (dr[0] <= grid_points)
-                & (grid_points <= dr[1])
-            )
-
-            index_list.append(keep_index)
-
-            new_grid_points.append(
-                grid_points[keep_index],
-            )
-
-        data_matrix = self.data_matrix[(slice(None),) + tuple(index_list)]
+        slice_list = []
+        for (a, b), dim_points in zip(domain_range, self.grid_points):
+            ia = np.searchsorted(dim_points, a)
+            ib = np.searchsorted(dim_points, b, 'right')
+            slice_list.append(slice(ia, ib))
+        grid_points = [g[s] for g, s in zip(self.grid_points, slice_list)]
+        data_matrix = self.data_matrix[(slice(None),) + tuple(slice_list)]
+
+        # Ensure that boundaries are in grid_points.
+        if with_bounds:
+            # Proceed dim by dim
+            for dim, (a, b) in enumerate(domain_range):
+                dim_points = grid_points[dim]
+                grid_points[dim] = []
+                add_low, add_high = False, False
+                if a < dim_points[0]:
+                    add_low = True
+                    grid_points[dim].append(a)
+                if b > dim_points[-1]:
+                    add_high = True
+                    grid_points[dim].append(b)
+                if not (add_low or add_high):
+                    grid_points[dim] = dim_points
+                    continue
+                # Interpolate on the edge.
+                eval_points = np.stack(
+                    np.meshgrid(*grid_points, indexing='ij'), axis=-1
+                )  # Shape (*map(len, grid_points), self.dim_domain)
+                eval_flat = eval_points.reshape(-1, self.dim_domain)
+                edge_values = self(eval_flat).reshape(
+                    len(self), *map(len, grid_points), self.dim_codomain)
+                # Add to self. Double if statement to only `insert` once.
+                if add_low:
+                    if add_high:
+                        data_matrix = np.concatenate(
+                            (edge_values.take([0], axis=dim+1),
+                             data_matrix,
+                             edge_values.take([1], axis=dim+1)),
+                            axis=dim+1)
+                        dim_points = np.concatenate(([a], dim_points, [b]))
+                    else:
+                        data_matrix = np.concatenate(
+                            (edge_values, data_matrix), axis=dim+1)
+                        dim_points = np.concatenate(([a], dim_points))
+                else:
+                    if add_high:
+                        data_matrix = np.concatenate(
+                            (data_matrix, edge_values), axis=dim+1)
+                        dim_points = np.concatenate((dim_points, [b]))
+                grid_points[dim] = dim_points
```

Comment:
This code is a bit difficult to maintain. Please check if this suggestion does the same:
```suggestion
            # Proceed dim by dim
            for dim, (a, b) in enumerate(domain_range):
                left = []
                right = []

                if a < grid_points[dim][0]:
                    left.append([a])

                if b > grid_points[dim][-1]:
                    right.append([b])

                grid_points[dim] = np.concatenate(left + [grid_points[dim]] + right)
            data_matrix = self(grid_points, grid=True)
```

Reply:
Oooh nice, I didn't know about the `grid` option of `FDataGrid.__call__`. This will certainly simplify things (as generating the evaluation mesh is no longer required).

On the other hand, I think your suggestion will reevaluate the function on *the whole domain* instead of on *the boundary*. If true, this is too costly to be considered and addition dimension by dimension should be preferred, imo. For example, with `dim_domain =3` and approximately `10` grid points per axis, your suggestion will evaluate `self` at `1000` points instead of only `60` being necessary.

I will incorporate the simplified syntax (with `grid` option) in the PR when I have the chance! I will also check how the `grid` option was implemented to ensure it is not slower (or not too much if at all) than the mesh approach.

Reply:
It is true that it evaluates all the points. However, I am not sure if it will be noticeably slower, as most part of the computation would be vectorized, if I am not mistaken. Unless it is noticeably slower I think that I prefer this version as it seems more readable to me (and this function is unlikely to be in a bottleneck anyway).